src/DataTypes/Serializations/ISerialization.h

```diff
@@ -152,6 +152,11 @@ class ISerialization : private boost::noncopyable, public std::enable_shared_fro
             ObjectStructure,
             ObjectData,
 
+            VariantDiscriminators,
+            VariantOffsets,
+            VariantElements,
```

Comment:
Because of the way how we create columns for subcolumns using `SubcolumnCreator`.
To create a proper subcolumn `variant.Type`, we need to use a `VariantSubcolumnCreator` on a nested sparse variant in function `ISerialization::createFromPath(...)`.

In `ISerialization::createFromPath(...)` we call `creator` for all elements in path that are before element that has subcolumn:
https://github.com/ClickHouse/ClickHouse/blob/61889faabd72dcf6199389df581e22d0c4cb8a8f/src/DataTypes/Serializations/ISerialization.cpp#L300-L314

Let's take an exaple: `Tuple(v Variant(String, UInt32), ...)` and we want to select subcolumn `.v.String`.
Without `VariantElements ` subcolumn `.v.String` will have path `TupleElement.VariantElement.Regular`. And when we want to create a column for subcolumn with path `TupleElement.VariantElement` the prefix will be `2`, so we will use data from `VariantElement` substream and use it in `SerializationNamed::SubcolumnCreator` from `TupleElement`. But we don't call `VariantSubcolumnCreator` stored in `VariantElement` stream.
So, I added `VariantElements` stream to store  `VariantSubcolumnCreator` in it, so we will have path `TupleElement.VariantElements.VariantElement` and when creating subcolumn for `VariantElement` we will use creator from substream `VariantElements`.

Actually, the same problem should happen for `Tuple` subcolumns, we actually won't use `SerializationNamed::SubcolumnCreator`, but it doesn't lead to problems as this is trivial creator.

Maybe we should change the logic of using these creators, WDYT?

Reply:
Ok, I understood, looks ok.